### PR TITLE
Refactor handle errors

### PR DESCRIPTION
**Pull request by frosty00**
there is still a lot of nesting to remove, but this is a start

### CODE REVIEW

File: js/exmo.js [hunk 1]
```diff
@@ -1024,12 +1024,9 @@ module.exports = class exmo extends Exchange {
     }
 
     handleErrors (httpCode, reason, url, method, headers, body, response) {
-        if (typeof body !== 'string')
-            return; // fallback to default error handler
-        if (body.length < 2)
+        if (typeof response === 'undefined')
```
Comment (kroitor) [inline js/exmo.js]: I suggest that we use a slightly shorter version without string-comparisons:

```JavaScript
if (response === undefined) {
    // ...
}
```

Can you plz add this edit everywhere below?

Comment (kroitor) [reply js/exmo.js]: > If you think the former is better I can blast the latter with a regex.
That would be great! ) If you will, please do it as a standalone PR though, just to keep things separate. Thx! )

File: js/kraken.js [hunk 2]
```diff
@@ -1174,7 +1174,6 @@ module.exports = class kraken extends Exchange {
         if (body.indexOf ('Invalid arguments:volume') >= 0)
             throw new InvalidOrder (this.id + ' ' + body);
         if (body[0] === '{') {
-            response = JSON.parse (body);
```
Comment (kroitor) [inline js/kraken.js]: There's an issue with this PR that might require a bit more work )

The reason why I'm not merging it and why I'm very suspicious on any PR that does a massive regex substitution is because I'm almost sure you haven't tested each of those exchanges after the substitution has been made ) It isn't really your fault, because the tests should have revealed the issue... 

This is a combination of a few details of Kraken with a design issue in the parsing chain:

- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/js/base/Exchange.js#L638
- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/js/base/Exchange.js#L639

The above two lines will only fill `json` if `parseJsonResponse` is true.
However, Kraken does not have an endpoint for order limits, so we scrape them from their web zendesk page:

- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/js/kraken.js#L226

For that reason, Kraken does not always respond JSON to us – can be HTML sometimes. Because of that, the following lines will set `parseJsonResponse` to false and will define parsing in the `request` method:

- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/js/kraken.js#L20
- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/js/kraken.js#L1194

So, removing the `response = JSON.parse (body)` from `handleErrors` in Kraken will effectively break the markets and the entire implementation of Kraken in this case.

The tests haven't revealed this issue because Kraken is skipped there (which is very unintuitive, I know):

- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/keys.json#L67

So, simply removing that line from `handleErrors` won't work. I'm going to enable Kraken back on in the tests to demonstrate the problem.

Comment (kroitor) [reply js/kraken.js]: To solve for this, I would propose to remove the `parseJsonResponse` flag completely and to always determine the type of response and decide whether JSON-parsing is needed based on the contents of the response itself. In other words we should only parse it if it is a JSON-encoded string, otherwise return a plaintext/HTML string – for all types of requests in all exchanges and all endpoints. 

Also, I would not try to parse JSON and catch the exception, instead, I think I would do basic checks for the first characters in the response before the parsing and would skip the parsing entirely for non-JSON responses.

Let me know what you think on the above.

Comment (frosty00) [reply js/kraken.js]: You said yourself that you don't like string comparisons, so I don't think iterating over each response is a viable solution, as it is O(n) and if the response had already been parsed it would make it O(2n). 

I'm not a huge fan of the `parseJsonResponse` flag either, because I'm not sure it works for asynchronous python. There is no point trying to do basic checks on the response in `handleErrors` anymore since currently we pass the parsed response to `handleErrors`, so we have a limited number of states:

* response is undefined and therefore the body cannot be parsed
* response is defined as a key-value table and it contains the parsed json response

Therefore if we are always expecting json we should do:

```
if (response === undefined)
    return
```

---

and if we need to do two different things depending on the response format then we should do

```
if (response === undefined)
     // do stuff with body text
else {
    // do stuff with json response
}
```

Comment (kroitor) [reply js/kraken.js]: > I'm not a huge fan of the `parseJsonResponse` flag either, because I'm not sure it works for asynchronous python.

Well, it doesn't, and it's one of the reasons I'd like to get rid of it.

> There is no point trying to do basic checks on the response in `handleErrors` anymore since currently we pass the parsed response to `handleErrors`, so we have a limited number of states:
> - response is undefined and therefore the body cannot be parsed
> - response is defined as a key-value table and it contains the parsed json response

Let me clarify

> There is no point trying to do basic checks on the response in `handleErrors` anymore

I was not suggesting  to do that in `handleErrors`, I meant to do that here universally for all exchanges, while keeping your approach to passing the `response` to `handleErrors`:

- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/js/base/Exchange.js#L638
- https://github.com/frosty00/ccxt/blob/refactor-handleErrors/js/base/Exchange.js#L639

In this particular case we have a third erroneous state that causes a bug and a break in this branch:

- `response is undefined because `parseJsonResponse` is set to `false`, but the body can be parsed – because `parseJsonResponse` is `false` all responses come into handleErrors as strings

So we will have to fix it, i suggest that we just do it right, and what I was talking about does not really contradict to what you said, I mean in order to fix it, we will have to either patch handleErrors for Kraken which is a compromise, or we could instead remove the `parseJsonResponse` flag completely from lines 638+639 and only attempt to parse JSON if JSON is there (not HTML), regardless of any flags. For that I would suggest checking the symbols instead of exception+catch. Does that clarify anything? )

Comment (frosty00) [reply js/kraken.js]: Well I have bad news, currently `parse_json` in python relies on `self.parseJsonResponse`:

https://github.com/ccxt/ccxt/blob/master/python/ccxt/base/exchange.py#L469

^ what is happening for kraken is really ugly. The json_response is set to the string of the html so it does not throw an error (because `self.parseJsonResponse` is set). We will have to refactor the base class to take this into account and remove all instances of `parseJsonResponse`. However, currently `handleRestResponse` in python also throws an error if the json_response is None. 

Basically it needs heavy refactoring in Python. I agree with the optimisation to not try and parse the json if the first character is not '{', but should we even bother since it is up to the JSON parser to handle these cases (it should stop iterating by itself if it comes across an invalid character).

Comment (kroitor) [reply js/kraken.js]: > Well I have bad news, currently parse_json in python relies on self.parseJsonResponse:

Not news to me )) The code that corresponds to `parse_json` in Python depends on self.parseJsonResponse everywhere now and that is something that we should fix I think. And yes, it will require some refactoring, but, tbh I don't think it should be a lot of refactoring. I'd rather do it gradually, by removing the `parseJsonResponse` first and adding the simple parsing logic that I've tried to explain above in all  base classes – that should not really involve more than a few lines in each baseclass.

> By symbols do you mean characters?

 Yep, that's correct, i mean characters...

> but should we even bother since it is up to the JSON parser to handle these cases

Well, I think comparing a couple of characters for equality is a much lighter task than calling the json parsing chain with multiple jumps over the stack and throwing the exception from there and catching the exception later just to come to the same conclusion, so that was more like a suggestion.
